Draft for JOSS publication (#190)
easystats/datawizard: paper/paper.Rmd

---

> title: "datawizard: An R Package for Easy Data Wrangling and Transformations"

Author: I am not fully sold on the terms I am using here: Data Wrangling and Data Transformations.

Basically, we should distinguish between two key categories in which _datawizard_ functions can be broken down:
one that deal with wrangling and the other that deal with statistical transformations. 

But I am not sure what other term to use here, or should we just give up on trying to force the section to have "data" prefix?

In that case, this would be: 
> datawizard: An R Package for Easy Data Wrangling and Statistical Transformations

> title: "datawizard: An R Package for Easy Data Wrangling and Statistical Transformations"

Reviewer: I would call it "Data Preparation", it encompasses wrangling + transfo + implies that it's pre-analysis

Author: Good point. Done!

tags:
  - R
  - easystats
authors:
- affiliation: 1
  name: Indrajeet Patil
  orcid: 0000-0003-1995-6531
- affiliation: 2
  name: Dominique Makowski
  orcid: 0000-0001-5375-9967
- affiliation: 3
  name: Mattan S. Ben-Shachar
  orcid: 0000-0002-4287-4801
- affiliation: 4
  name: Brenton M. Wiernik
  orcid: 0000-0001-9560-6336
- affiliation: 5
  name: Etienne Bacher
  orcid: 0000-0002-9271-5075 
- affiliation: 6
  name: Daniel Lüdecke
  orcid: 0000-0002-8895-3206
  
affiliations:
- index: 1
  name: esqLABS GmbH, Germany
- index: 2
  name: Nanyang Technological University, Singapore
- index: 3
  name: Ben-Gurion University of the Negev, Israel
- index: 4

>   name: Facebook

Reviewer 2: ```suggestion
  name: Independent Researcher
```

Reviewer 2: Add an author note for me: "Brenton Wiernik is currently an independent researcher and Research Scientist at Meta,  Demography and Survey Science. The current research was in an independent capacity."

  orcid: 0000-0001-9560-6336
- affiliation: 5
  name: Etienne Bacher

>   orcid: 0000-0002-9271-5075 

Author: Can you please also add this to `DESCRIPTION` file when you make a PR for #188

Reviewer 3: yes I will

In addition to providing functions to clean messy data, `{datawizard}` also provides helpers for the other important step of data analysis: transforming the cleaned data further for setting up statistical models. For example, one may need to standardize certain variables, normalize range of some variables, adjust the data for effect of some variables, etc.

Lastly, `{datawizard}` also provides a toolbox to create a detailed profile of data properties.

# Features

## Data wrangling

The raw data is rarely in a state that it can be directly fed into a statistical model. It often needs to be modified in various ways. For example, columns need to be renamed and/or reordered, data scattered across multiple tables needs to be joined, certain parts of the data need to be left out, etc. 

`{datawizard}` provides various functions for cleaning and preparing data (see Table 1).

Function           | Operation                             |
------------------ | --------------------------------------|
`data_filter()`    | to select only certain *observations* |
`data_select()`    | to select only a few *attributes*     |

> `data_extract()`   | to extract a single *attribute*       |

Author: More common in DS literature, but, you are right, _variable_ is better.

`data_rename()`    | to rename attributes                  |
`reshape_longer()` | to convert data from wide to long     |
`reshape_wider()`  | to convert data from long to wide     |
`data_join()`      | to join two data frames               |
    ...            |        ...                            |

> Table: The table below lists a few key functions offered by *datawizard* for data wrangling. To see the full list, see the package website: <https://easystats.github.io/datawizard/>

Author: I am deliberately not being comprehensive because we might add more functions in the future and `...` cover all of those.

We will look at one example function that converts data in wide format to tidy/long format:

```{r}
stocks <- data.frame(
  time = as.Date('2009-01-01') + 0:4,
  X = rnorm(5, 0, 1),
  Y = rnorm(5, 0, 2)
)

stocks

data_to_long(
  stocks,
  select = -c("time"),
  colnames_to = "stock",
  values_to = "price"
)
```

## Data transformations

Even after getting the raw data in the needed format, we may further need to transform certain variables further to meet requirements imposed by the statistical model.

`{datawizard}` provides a rich collection of such functions for transforming variables (see Table 2).

Function           | Operation                                     |
------------------ | ----------------------------------------------|
`standardize()`    | to center and scale data                      |
`normalize()`      | to scale variables to 0-1 range               |
`adjust()`         | to adjust data for effect of other variables  |
`data_shift()`     | to shift numeric value range                  |
`ranktransform()`  | to convert numeric values to integer ranks    |
    ...            |        ...                                    |

> Table: The table below lists a few key functions offered by *datawizard* for data transformations. To see the full list, see the package website: <https://easystats.github.io/datawizard/>

Author: I am deliberately not being comprehensive because we might add more functions in the future and `...` cover all of those.

We will look at one example function that standardizes (i.e. centers and scales) data so that it can be expressed in terms of standard deviation:

```{r}
d <- data.frame(
  a = c(-2, -1, 0, 1, 2),
  b = c(3, 4, 5, 6, 7)
)

standardize(d, center = c(3, 4), scale = c(2, 4))
```

## Data properties

> The workhorse function to get a comprehensive summary of data properties is `describe_distribution()`, which combines a set of indices (e.g., measures of centrality, dispersion, range, skewness, kurtosis, etc.) computed by other functions in `{datawizard}`.

Author: Not sure if a table is necessary here. We don't have many functions.

```{r}
describe_distribution(mtcars$wt)

> ```

Reviewer 3: add an example to show that it also works with a full dataset?

Author: Done.

But the table output overflows in PDF; we will need to deal with that somehow.

# Licensing and Availability

> *datawizard* is licensed under the GNU General Public License (v3.0), with all source code openly developed and stored at GitHub (<https://github.com/easystats/datawizard>), along with a corresponding issue tracker for bug reporting and feature enhancements. In the spirit of honest and open science, we encourage requests, tips for fixes, feature updates, as well as general questions and concerns via direct interaction with contributors and developers.

Reviewer 3: we need to add a license to the repo

Author: Done.

knitr::opts_chunk$set(
  collapse = TRUE,
  out.width = "100%",
  dpi = 300,
  comment = "#>",
  message = FALSE,
  warning = FALSE
)

> library(datawizard)

Reviewer 2: ```suggestion
library(datawizard)
set.seed(2016)
```

```

# Summary

> The `{datawizard}` package for the R programming language [@base2021] provides a lightweight toolbox to assist in keys steps involved in any data analysis workflow: (1) wrangling the raw data to get it in the needed form, (2) applying preprocessing steps and statistical transformations, and (3) conducting reasonableness checks to ensure transformed data are high quality. Therefore, it can be a valuable tool for R users and developers looking for a lightweight option for data preparation.

Reviewer 2: ```suggestion
The `{datawizard}` package for the R programming language [@base2021] provides a lightweight toolbox to assist in keys steps involved in any data analysis workflow: (1) wrangling the raw data to get it in the needed form, (2) applying preprocessing steps and statistical transformations, and (3) compute statistical summaries of data properties and distributions. Therefore, it can be a valuable tool for R users and developers looking for a lightweight option for data preparation.
```

# Statement of Need

The `{datawizard}` package is part of `{easystats}`, a collection of R packages designed to make statistical analysis easier (@Ben-Shachar2020, @Lüdecke2020parameters, @Lüdecke2020performance, @Lüdecke2021see, @Lüdecke2019, @Makowski2019, @Makowski2020). As this ecosystem follows a "0-external-hard-dependency" policy, a base R data manipulation package that relies only on base R needed to be created. In effect, `{datawizard}` provides data processing backend for this entire ecosystem. 
In addition to its usefulness to the `{easystats}` ecosystem, it also provides *an* option for R users and package developers if they wish to keep their (recursive) dependency weight to a minimum (for other options, see @Dowle2021, @Eastwood2021, etc.).

> Because `{datawizard}` is also meant to be used and adopted easily by a wide range of users, its workflow and syntax are designed to be similar to `{tidyverse}` (@Wickham2019), a widely used ecosystem of R packages. Thus, users familiar with the `{tidyverse}` can easily translate their knowledge and make full usage of `{datawizard}`.

Reviewer 2: ```suggestion
Because `{datawizard}` is also meant to be used and adopted easily by a wide range of users, its workflow and syntax are designed to be similar to `{tidyverse}` (@Wickham2019), a widely used ecosystem of R packages. Thus, users familiar with the `{tidyverse}` can easily translate their knowledge and make full use of `{datawizard}`.
```

In addition to being a lightweight solution to clean messy data, `{datawizard}` also provides helpers for the other important step of data analysis: applying statistical transformations to the cleaned data while setting up statistical models. This includes various types of data standardization, normalization, rank-transformation, and adjustment. These transformations, although widely used, are not currently collectively implemented in a package in the R ecosystem, so `{datawizard}` can help new R users in finding the transformation they need.

> Lastly, `{datawizard}` also provides a toolbox to create a detailed profile of data properties. This is a common step in data analysis, but it is not available in base R or many modeling packages, so its inclusion makes `{datawizard}` a one-stop-shop for data preparation tasks.

Reviewer 2: ```suggestion
Lastly, `{datawizard}` also provides a toolbox to create a detailed profile of data properties. This is a common step in data analysis, but it is not readily available in base R or many modeling packages, so its inclusion makes `{datawizard}` a one-stop-shop for data preparation tasks.
```

> Lastly, `{datawizard}` also provides a toolbox to create a detailed profile of data properties. This is a common step in data analysis, but it is not available in base R or many modeling packages, so its inclusion makes `{datawizard}` a one-stop-shop for data preparation tasks.

Reviewer 2: ```suggestion
Lastly, `{datawizard}` also provides a toolbox to create a detailed summaries of data properties and distributions (e.g., tables of descriptive statistics for each variable). This is a common step in data analysis, but it is not available in base R or many modeling packages, so its inclusion makes `{datawizard}` a one-stop-shop for data preparation tasks.
```

# Features

## Data Preparation

> The raw data is rarely in a state that it can be directly fed into a statistical model. It often needs to be modified in various ways. For example, columns need to be renamed, certain portions of the data need to be filtered out, reshape data, data scattered across multiple tables needs to be joined, etc. 

Reviewer 3: "reshape data" -> "some columns need to be reshaped"?

`{datawizard}` provides a rich collection of such functions for transforming variables (see Table 2).

| Function          | Operation                                    |
| :---------------- | :------------------------------------------- |
| `standardize()`   | to center and scale data                     |
| `normalize()`     | to scale variables to 0-1 range              |
| `adjust()`        | to adjust data for effect of other variables |
| `slide()`         | to shift numeric value range                 |
| `ranktransform()` | to convert numeric values to integer ranks   |
| ...               | ...                                          |

Table: The table below lists a few key functions offered by *datawizard* for data transformations. To see the full list, see the package website: <https://easystats.github.io/datawizard/>

We will look at one example function that standardizes (i.e. centers and scales) data so that it can be expressed in terms of standard deviation:

```{r}
d <- data.frame(
  a = c(-2, -1, 0, 1, 2),
  b = c(3, 4, 5, 6, 7)
)

standardize(d, center = c(3, 4), scale = c(2, 4))
```

> ## Data Properties

Reviewer 2: ```suggestion
## Summaries of Data Properties and Distributions
```

The workhorse function to get a comprehensive summary of data properties is `describe_distribution()`, which combines a set of indices (e.g., measures of centrality, dispersion, range, skewness, kurtosis, etc.) computed by other functions in `{datawizard}`.

```{r eval=FALSE}
describe_distribution(mtcars)
```

```{r echo=FALSE, eval=TRUE, results="asis"}
library(kableExtra)
options(digits = 3)
kbl(describe_distribution(mtcars), format = "latex", booktabs = TRUE, linesep = "") |>
  kable_styling(latex_options = "scale_down")
```

# Licensing and Availability

> `{datawizard}` is licensed under the GNU General Public License (v3.0), with all source code openly developed and stored at GitHub (<https://github.com/easystats/datawizard>), along with a corresponding issue tracker for bug reporting and feature enhancements. In the spirit of honest and open science, we encourage requests, tips for fixes, feature updates, as well as general questions and concerns via direct interaction with contributors and developers.

Reviewer 3: "at Github" -> "on Github"?

# Acknowledgments

> `{datawizard}` is part of the collaborative [*easystats*](https://github.com/easystats/easystats) ecosystem. Thus, we thank the [members of easystats](https://github.com/orgs/easystats/people) as well as the users.

Reviewer 3: maybe link to https://easystats.github.io/easystats/ instead

In addition to its usefulness to the `{easystats}` ecosystem, it also provides *an* option for R users and package developers if they wish to keep their (recursive) dependency weight to a minimum (for other options, see @Dowle2021, @Eastwood2021, etc.).

Because `{datawizard}` is also meant to be used and adopted easily by a wide range of users, its workflow and syntax are designed to be similar to `{tidyverse}` (@Wickham2019), a widely used ecosystem of R packages. Thus, users familiar with the `{tidyverse}` can easily translate their knowledge and make full use of `{datawizard}`.

In addition to being a lightweight solution to clean messy data, `{datawizard}` also provides helpers for the other important step of data analysis: applying statistical transformations to the cleaned data while setting up statistical models. This includes various types of data standardization, normalization, rank-transformation, and adjustment. These transformations, although widely used, are not currently collectively implemented in a package in the R ecosystem, so `{datawizard}` can help new R users in finding the transformation they need.

> Lastly, `{datawizard}` also provides a toolbox to create a detailed profile of data properties. This is a common step in data analysis, but it is not readily available in base R or many modeling packages, so its inclusion makes `{datawizard}` a one-stop-shop for data preparation tasks.

Reviewer 2: ```suggestion
Lastly, `{datawizard}` also provides a toolbox to create detailed summaries of data properties and distributions (e.g., tables of descriptive statistics for each variable). This is a common step in data analysis, but it is not available in base R or many modeling packages, so its inclusion makes `{datawizard}` a one-stop-shop for data preparation tasks.
```
